src/expr/symbol_table.cpp

```diff
@@ -41,23 +42,165 @@ using ::std::pair;
 using ::std::string;
 using ::std::vector;
 
+// This data structure stores a trie of expressions with
+// the same name, and must be distinguished by their argument types.
+// It is context-dependent.
+class OverloadedTypeTrie
+{
+public:
+  OverloadedTypeTrie(Context * c ) :
+    d_overloaded_symbols(new (true) CDHashSet<Expr, ExprHashFunction>(c)) {
+  }  
+  ~OverloadedTypeTrie() {
+    d_overloaded_symbols->deleteSelf();
+  }
+  /** is this function overloaded? */
+  bool isOverloadedFunction(Expr fun) const;
+  
+  /** Get overloaded constant for type.
+   * If possible, it returns a defined symbol with name
+   * that has type t. Otherwise returns null expression.
+  */
+  Expr getOverloadedConstantForType(const std::string& name, Type t) const;
+  
+  /**
+   * If possible, returns a defined function for a name
+   * and a vector of expected argument types. Otherwise returns
+   * null expression.
+   */
+  Expr getOverloadedFunctionForTypes(const std::string& name, const std::vector< Type >& argTypes) const;
+  /** called when obj is bound to name, and prev_bound_obj was already bound to name */
+  void bind(const string& name, Expr prev_bound_obj, Expr obj);
+private:
+  /** Marks expression obj with name as overloaded. 
+   * Adds relevant information to the type arg trie data structure.
+  */
+  void markOverloaded(const string& name, Expr obj);
+  /** the null expression */
+  Expr d_nullExpr;
+  // The (context-independent) trie storing that maps expected argument
+  // vectors to symbols. All expressions stored in d_symbols are only 
+  // interpreted as active if they also appear in the context-dependent
+  // set d_overloaded_symbols.
+  class TypeArgTrie {
+  public:
+    // children of this node
+    std::map< Type, TypeArgTrie > d_children;
+    // symbols at this node
+    std::map< Type, Expr > d_symbols;
+  };
+  /** for each string with operator overloading, this stores the data structure above. */
+  std::unordered_map< std::string, TypeArgTrie > d_overload_type_arg_trie;
+  /** The set of overloaded symbols. */
+  CDHashSet<Expr, ExprHashFunction>* d_overloaded_symbols;
+};
+
+bool OverloadedTypeTrie::isOverloadedFunction(Expr fun) const {
+  return d_overloaded_symbols->find(fun)!=d_overloaded_symbols->end();
+}
+
+Expr OverloadedTypeTrie::getOverloadedConstantForType(const std::string& name, Type t) const {
```

Comment:
If I understand correctly, this just returns the first symbol, even if there is more than one. Is that right? Even if it's not in d_overloaded_symbols? This is different from getOverloadedFunctionForTypes ... why?  And can you update the comments for this function (also the SymbolTable function of the same name) to say how it behaves if there is more than one.

Reply:
Yes, you bring up a really strange case, the question is what do we do if the user inputs:

(declare-fun x () Int)
(declare-fun x () Int)

Technically the constants are indistinguishable so we can simply ignore the second.

I've updated this in my latest commit.

Reply:
Btw, there can't be more than one constant of a given type stored in this map.
The difference is we are querying by the argument types in the function case, and the range type in the constant case.
The function case can have multiple symbols since we can have ( T1 x ... x Tn ) -> U1 and  ( T1 x ... x Tn ) -> U2, but constants are all at the top-level in the trie since their arguments are different, thus there is no way to store two constants of the same type.